### PR TITLE
Request body examples

### LINKED ISSUE #380: Request examples
Is there a way or fix to add request examples like you can do in Swagger eg.

      requestBody:
        content:
          application/json:     # Media type
            schema:             # Request body contents
              $ref: '#/components/schemas/User'   # Reference to an object
            examples:    # Child of media type
              Jessica:   # Example 1
                value:
                  id: 10
                  name: Jessica Smith
              Ron:       # Example 2
                value:
                  id: 11
                  name: Ron Stewart

### PR DESCRIPTION
**Pull request by romanblanco**
## Problem
Closes https://github.com/rswag/rswag/issues/380

## Solution
Cherry-picking @kriston-costa's solution from https://github.com/instacart/rswag/pull/4

### This concerns this parts of the Open API Specification:
* [Request Body Examples](https://spec.openapis.org/oas/v3.1.0#request-body-examples)

### The changes I made are compatible with:
- [x] OAS2
- [x] OAS3
- [x] OAS3.1

### Related Issues

- https://github.com/rswag/rswag/issues/380
- https://github.com/instacart/rswag/pull/4/files

### Checklist
- [x] Added tests
- [x] Changelog updated
- [x] Added documentation to README.md

### Steps to Test or Reproduce

```ruby
# spec/integration/blogs_spec.rb
describe 'Blogs API' do
  path '/blogs/{blog_id}' do
    get 'Retrieves a blog' do
      request_example value: { some_field: 'Foo' }, name: 'request_example_1', summary: 'A request example'
      response 200, 'blog found' do
        ...
```

### CODE REVIEW

**Comment by kstevens715** [comment]
We need this exact feature. I tested this branch out in our project and it worked out great. For us, we wanted to be able to use the actual request from the spec as the example, so we did it in an after block. But I think what you have here is good enough for our use case. I'd love to see this get merged! Thanks for your work!

```ruby
  config.after(:each, operation: true, use_as_request_example: true) do |spec|
    spec.metadata[:operation][:request_examples] ||= []

    example = {
      value: JSON.parse(request.body.string, symbolize_names: true),
      name: 'request_example_1',
      summary: 'A request example'
    }

    spec.metadata[:operation][:request_examples] << example
  end
```

File: README.md [hunk 1]
```diff
@@ -144,6 +145,7 @@ There is also a generator which can help get you started `rails generate rspec:s
           tags 'Blogs', 'Another Tag'
           produces 'application/json', 'application/xml'
           parameter name: :id, in: :path, type: :string
+          request_example value: { some_field: 'Foo' }, name: 'basic', summary: 'Request example description'
```
Comment (bblimke) [inline README.md:148]: The method is `request_body_example` and not `request_example` so either README has wrong name or the method. Not sure what was the intention. `request_body_example` is probably more intuitive.

Comment (romanblanco) [reply README.md:148]: Thanks @bblimke, you're right!

https://github.com/rswag/rswag/pull/566

**Comment by harlantwood** [comment]
> We need this exact feature. I tested this branch out in our project and it worked out great. For us, we wanted to be able to use the actual request from the spec as the example, so we did it in an after block. But I think what you have here is good enough for our use case. I'd love to see this get merged! Thanks for your work!
> 
> ```ruby
>   config.after(:each, operation: true, use_as_request_example: true) do |spec|
>     spec.metadata[:operation][:request_examples] ||= []
> 
>     example = {
>       value: JSON.parse(request.body.string, symbolize_names: true),
>       name: 'request_example_1',
>       summary: 'A request example'
>     }
> 
>     spec.metadata[:operation][:request_examples] << example
>   end
> ```

This is awesome 😄 
To spell it out a bit more for people who need it (like me):

1) put this in your `swagger_helper.rb`
2) tag the spec that you want to use as a request example with `use_as_request_example`, eg:

```rb
      response(200, "successful", use_as_request_example: true) do
```

**Comment by harlantwood** [comment]
Since this PR is merged but not yet in a gem release, I am using this feature by loading the gem from github with the following entry in my `Gemfile`:

```rb
  gem "rswag-specs", git: "https://github.com/rswag/rswag.git", branch: "master", 
    glob: "rswag-specs/rswag-specs.gemspec" # to get request examples, merged in https://github.com/rswag/rswag/pull/555
```

**Comment by BenKanouse** [comment]
For my project I edited the above example to look like this:

``` ruby
  config.after(:each, operation: true, use_as_request_example: true) do |spec|
    spec.metadata[:operation][:request_examples] ||= []

    example = {
      value: JSON.parse(request.body.string, symbolize_names: true),
      name: spec.metadata[:response][:description].parameterize.underscore,
      summary: spec.metadata[:response][:description]
    }

    spec.metadata[:operation][:request_examples] << example
  end
  ```
  
  The main difference is that the name and summary of the example come from the spec's description. This means that you can have multiple `response` blocks use the `use_as_request_example` without them overwriting each-other and the name/summary in the ui looks a bit nicer than a static value.

**Pull request merged**